javascript/javascript1/week3/CactusIO-interactive.js

```diff
@@ -0,0 +1,82 @@
+let activities =[];
+let limit = 180;
+let value = 0;
+var options = { weekday: 'long', year: 'numeric', month: 'long', day: 'numeric' };
+var today  = new Date();
+
+function addMoreLimit(minutes){ // I added the "addMoreLimit" function if user wants to change the limit
+    if(typeof(minutes) != "number" && minutes < 0){
+        return "Error! Incorrect number";
+    } else {
+        if(limit < 300){
+            if(minutes < 120){
+                limit += minutes;
+                console.log(`${minutes} minutes added. Now limit is ${limit} minutes`);
+            } else{
+                return "It's too much. Add less";
+            }
+        } else {
+            return "You're on the phone a lot today!";
+        }
+    }
+}
+
+function addActivity(date, activity, duration){
+    x = ((typeof(activity) === "string" && activity != "") && (typeof(date) === "string" && date != "") && typeof(duration) === "number");
+    if(x){
+        if(limit < value+duration){
+            return console.log("You have reached your limit, no more smartphoning for you!");
+        } else{
+            value =+ duration;
+            return activities.push({date, activity, duration});
+        }
+    } else {
+        return "Error";
+    }
+}
+
+function showStatus(activities){
+    if(activities.length === 0){
+        return "Add some activities before calling showStatus";
+    }
+
+    let count = 0;
+    let amount = 0;
+    for (let i = 0; i < activities.length; i++) {
+        count += 1;
+        amount += activities[i].duration;
+    }
+
+    console.log(`You have added ${count} activities. They amount to ${amount} min. to usage`);
+    if(amount >= limit){
+        console.log("You have reached your limit, no more smartphoning for you!");
+    }
+}
+
+function mostTimeAct(){
+    if(activities.length>0){
+        let mostAct = activities[0];
+        for (let i = 0; i < activities.length; i++) {
+            if(activities[i].duration > mostAct.duration){
+                mostAct = activities[i];
+            }
+        }
+        return `The most time activitie: ${mostAct.activity}. It takes ${mostAct.duration} minutes`;
+    } else {
+        return "0 activities. Add few more."
+    }
+}
+
+addActivity('2023, 7 , 19', "Youtube", 30);
+showStatus(activities);
+addActivity('2023, 7 , 18', "Twitch", 60);
+showStatus(activities);
+addActivity('2023, 7 , 18', "CS2", 180);
```

Comment:
If I split this 180 minutes activity into two activities of 90 minutes each, I do not get an error stating I have exceeded my daily limit. 🐛 Can you see what goes wrong? (Hint: something sneaky is going on in line 30!)

Having the variable `value`, which I assume is supposed to contain the sum of activities, makes the code a bit fragile. First and foremost, we are now storing duplicate data, i.e., `value` is computed based on what is in `activities`. This means they can get out-of-sync! 😱 I would avoid this variable and simply compute the sum when it is needed; it is fine four our purposes here! (Maybe introduce a function that does this so it can be reused!)

Generally, I would recommend limiting global variables, i.e., variables not inside functions, to a minimum as it can be hard to maintain code that has too many globals. Generally, you should strive to have functions only be determined by their inputs; if you want to read more about this try googling "side effects". 😄